### PR TITLE
rgw: pass authentication domain to civetweb

### PR DESCRIPTION
**Pull request by theanalyst**
This is needed to support absolute uris, as civetweb further cross
checks whether the request is made to the same domain or not, otherwise
civetweb ends up sending NULL response back to the client. Also set the
REQUEST_URI to the full request uri, since the old uri parameter is
deprecated.

This needs https://github.com/ceph/civetweb/pull/16 also merged for full support of subdomain style requests

Fixes: http://tracker.ceph.com/issues/17657
Signed-off-by: Abhishek Lekshmanan <abhishek@suse.com>

### CODE REVIEW

**Comment by theanalyst** [comment]
civetweb offers the choice of both local_uri and the request_uri (which is currently used), if we set the uri as local_uri we could avoid an additional parsing we do in RGW where we get the path from abs uri, but I guess it is better to pass on the full request unmodified by civetweb, so set the value to request_uri

**Comment by theanalyst** [comment]
Tested initially with aws-sdk-go, which by default creates requests with abs. urls,  eg. https://gist.github.com/theanalyst/45732de392237ac3e01c26bc4ff9d19f 

though by default boto doesn't allow this sort of urls, we can override boto's parent libraries to make similar requests, which we can probably use to add to s3 tests, an example here https://gist.github.com/theanalyst/ac32ed95a96f36d5ee10efc2673c6b7d

File: src/rgw/rgw_civetweb_frontend.cc [hunk 1]
```diff
@@ -64,6 +64,9 @@ int RGWCivetWebFrontend::run()
     conf_map["run_as_user"] = std::move(uid_string);
   }
 
+  /* Set authentication domains, this is needed for handling absolute uris in
+     requests */
+  conf_map["authentication_domain"] = g_conf->rgw_dns_name;
```
Comment (yehudasa) [inline src/rgw/rgw_civetweb_frontend.cc]: @theanalyst this is not necessarily the only domain that we support. We'd have a different domain for static websites, and nowadays we check every domain to match buckets (see code in rgw_rest.cc). Instead of having civetweb limit us to a specific domain, I'd rather have it send us everything and we can make the decision in rgw.

Comment (theanalyst) [inline src/rgw/rgw_civetweb_frontend.cc]: currently the problem is that civetweb validates that the request is made to the same domain, or errors out there we could probably modify civetweb to not do any checking at all on domain and only rely on our checks,

Comment (theanalyst) [inline src/rgw/rgw_civetweb_frontend.cc]: @yehudasa this is the current behaviour of civetweb (with the proposed cherry-pick) https://github.com/civetweb/civetweb/issues/197#issuecomment-255186417

Comment (yehudasa) [inline src/rgw/rgw_civetweb_frontend.cc]: Yeah, so we should modify civetweb (add a new configurable to allow that so that it can be accepted upstream).

Comment (theanalyst) [inline src/rgw/rgw_civetweb_frontend.cc]: ack, will add a configurable to bypass domain check

**Comment by theanalyst** [comment]
@yehudasa updated the civetweb pr to use a new configurable to disable auth domain check, and updated our configuration of civetweb. I'll open a new pr against civetweb upstream for the configurable as well.

**Comment by theanalyst** [comment]
@yehudasa updated the civetweb pr https://github.com/ceph/civetweb/pull/16, also upstream merged the https://github.com/civetweb/civetweb/pull/397

**Comment by theanalyst** [comment]
@yehudasa can you take a look

File: src/rgw/rgw_civetweb.cc [hunk 2]
```diff
@@ -111,7 +111,7 @@ void RGWCivetWeb::init_env(CephContext *cct)
   }
 
   env.set("REQUEST_METHOD", info->request_method);
-  env.set("REQUEST_URI", info->uri);
+  env.set("REQUEST_URI", info->request_uri); // get the full uri, we anyway handle abs uris later
```
Comment (yehudasa) [inline src/rgw/rgw_civetweb.cc]: @theanalyst is this change needed? it's not clear to me what's the effect of this line

**Comment by yehudasa** [comment]
@theanalyst need to squash commits

File: src/rgw/rgw_civetweb.cc (hunk 2 above)
Comment (theanalyst) [inline src/rgw/rgw_civetweb.cc]: https://github.com/civetweb/civetweb/blob/59fa55a87cdc8397f4f88e4b76881857f6e0622e/docs/api/mg_request_info.md civetweb marked their uri field as deprecated and introduced local_uri and request_uri, and uri field was set as local_uri for now, but maybe removed in future

**Comment by theanalyst** [comment]
squashed the commits, we would want https://github.com/ceph/civetweb/pull/16 merged and then updatin the submodule ref before the commit goes in right?

**Comment by yehudasa** [comment]
@theanalyst yes, we need to have it merged. Submodule ref change needs to be part of this PR.

**Comment by theanalyst** [comment]
@yehudasa is https://github.com/ceph/civetweb/pull/16 ok for merging?

**Comment by theanalyst** [comment]
@yehudasa ping

**Comment by yehudasa** [comment]
@theanalyst this is still missing the submodule update though

**Comment by theanalyst** [comment]
@yehudasa adding, didn't know what hash to set it to before, setting to de23828cea740032428ce6f288ad061b429ce68e

**Comment by theanalyst** [comment]
scheduled a small subset of s3 suites for both before and after this PR (luminous branch and the current branch) 
http://pulpito.ceph.com/abhi-2017-05-17_16:13:38-rgw-wip-absolute-urls---basic-smithi/
http://pulpito.ceph.com/abhi-2017-05-17_16:12:50-rgw-wip-luminous---basic-smithi/

**Comment by theanalyst** [comment]
@yehudasa I have a small QE run which is green for s3tests before this pr except for absolute urls, is this sufficient or do we need a full QE run with a subset or so?

**Comment by theanalyst** [comment]
scheduled a 1/3 subset at http://pulpito.ceph.com/abhi-2017-05-18_09:58:15-rgw-wip-absolute-urls---basic-smithi/

**Comment by theanalyst** [comment]
there were failures in a few s3tests and swift tests, but these were headers/url encoding tests failing on apache fastcgi, maybe relevant to call teuthology with a `--filter-out=apache`

**Comment by cbodley** [comment]
> there were failures in a few s3tests and swift tests

confirmed that the test failures were known issues with apache frontend 👍

**Comment by theanalyst** [comment]
@yehudasa can you take a look again?
